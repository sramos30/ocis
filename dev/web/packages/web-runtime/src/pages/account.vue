<template>
  <main id="account" class="oc-height-1-1 oc-m">
    <div class="oc-flex oc-flex-between oc-flex-bottom oc-width-1-1 oc-border-b oc-py">
      <h1 id="account-page-title" class="oc-page-title oc-m-rm">{{ pageTitle }}</h1>
      <div>
        <oc-button
          v-if="editUrl"
          variation="primary"
          type="a"
          :href="editUrl"
          data-testid="account-page-edit-url-btn"
        >
          <oc-icon name="edit" />
          <translate>Edit</translate>
        </oc-button>
        <oc-button
          v-else-if="editRoute"
          variation="primary"
          type="router-link"
          :to="editRoute"
          data-testid="account-page-edit-route-btn"
        >
          <oc-icon name="edit" />
          <translate>Edit</translate>
        </oc-button>
      </div>
    </div>
    <h2 v-translate class="oc-text-bold oc-text-initial oc-mb">Account Information</h2>
    <dl class="account-page-info oc-flex oc-flex-wrap">
      <div class="account-page-info-username oc-mb oc-width-1-2@s">
        <dt v-translate class="oc-text-normal oc-text-muted">Username</dt>
        <dd>
          {{ user.username || user.id }}
        </dd>
      </div>
      <div v-if="user.username && user.id" class="account-page-info-userid">
        <dt v-translate class="oc-text-normal oc-text-muted">User ID</dt>
        <dd>
          {{ user.id }}
        </dd>
      </div>
      <div class="account-page-info-displayname oc-mb oc-width-1-2@s">
        <dt v-translate class="oc-text-normal oc-text-muted">Display name</dt>
        <dd>
          {{ user.displayname }}
        </dd>
      </div>
      <div class="account-page-info-email oc-mb oc-width-1-2@s">
        <dt v-translate class="oc-text-normal oc-text-muted">Email</dt>
        <dd>
          <template v-if="user.email">{{ user.email }}</template>
          <span v-else v-translate>No email has been set up</span>
        </dd>
      </div>
      <div class="account-page-info-groups oc-mb oc-width-1-2@s">
        <dt v-translate class="oc-text-normal oc-text-muted" @click="loadGroups">
          Group memberships
        </dt>
        <dd data-testid="group-names">
          <oc-spinner
            v-if="loadingGroups"
            :aria-label="$gettext('Loading group membership information')"
          />
          <span v-else-if="groupNames">{{ groupNames }}</span>
          <span v-else v-translate data-testid="group-names-empty"
            >You are not part of any group</span
          >
        </dd>
      </div>
    </dl>
  </main>
</template>

<script>
import { mapGetters } from 'vuex'
export default {
  name: 'Personal',
  data() {
    return {
      loadingGroups: true,
      groups: []
    }
  },
  computed: {
    ...mapGetters(['user', 'configuration', 'getNavItemsByExtension', 'apps']),
    isAccountEditingEnabled() {
      return !this.apps.settings
    },
    pageTitle() {
      return this.$gettext(this.$route.meta.title)
    },
    editUrl() {
      if (!this.isAccountEditingEnabled) {
        return null
      }
      return this.configuration.server.replace(/\/$/, '') + '/index.php/settings/personal'
    },
    editRoute() {
      const navItems = this.getNavItemsByExtension('settings')
      if (navItems.length > 0) {
        return navItems[0].route || {}
      }
      return null
    },
    groupNames() {
      return this.groups.join(', ')
    }
  },
  mounted() {
    this.loadGroups()
  },
  methods: {
    async loadGroups() {
      this.groups = await this.$client.users.getUserGroups(this.user.id)
      this.loadingGroups = false
    }
  }
}
</script>
<style lang="scss">
.account-page-info dd {
  margin-left: 0;
}
</style>
