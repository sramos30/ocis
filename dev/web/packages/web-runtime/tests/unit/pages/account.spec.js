import account from '../../../src/pages/account.vue'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import { createStore } from 'vuex-extensions'
import GetTextPlugin from 'vue-gettext'
import VueCompositionAPI from '@vue/composition-api'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueCompositionAPI)

localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

localVue.directive('translate', {
  inserted: (el) => {}
})

const $route = {
  meta: {
    title: 'Some Title'
  }
}

const selectors = {
  pageTitle: '.oc-page-title',
  loaderStub: 'oc-spinner-stub',
  editUrlButton: '[data-testid="account-page-edit-url-btn"]',
  editRouteButton: '[data-testid="account-page-edit-route-btn"]',
  accountPageInfo: '.account-page-info',
  groupNames: '[data-testid="group-names"]',
  groupNamesEmpty: '[data-testid="group-names-empty"]'
}

describe('account page', () => {
  describe('header section', () => {
    it('renders page title', () => {
      const wrapper = getWrapper()
      const pageTitle = wrapper.find(selectors.pageTitle)
      expect(pageTitle.exists()).toBeTruthy()
      expect(pageTitle.text()).toBe($route.meta.title)
    })

    describe('edit buttons', () => {
      describe('edit url button', () => {
        it('should be displayed if not running with ocis', async () => {
          const store = getStore({
            server: 'http://server/address/',
            isAccountEditingEnabled: true
          })
          const wrapper = getWrapper(store)
          await wrapper.setData({ loadingGroups: false })
          const editUrlButton = wrapper.find(selectors.editUrlButton)
          const editRouteButton = wrapper.find(selectors.editRouteButton)
          expect(editUrlButton).toMatchSnapshot()
          expect(editRouteButton.exists()).toBeFalsy()
        })
        it('should not be displayed if running with ocis', async () => {
          const store = getStore({
            server: 'http://server/address/',
            isAccountEditingEnabled: false
          })
          const wrapper = getWrapper(store)
          await wrapper.setData({ loadingGroups: false })
          const editUrlButton = wrapper.find(selectors.editUrlButton)
          expect(editUrlButton.exists()).toBeFalsy()
        })
      })
      describe('edit route button', () => {
        it('should be displayed if running with ocis and has navItems', async () => {
          const store = getStore({
            isAccountEditingEnabled: false,
            getNavItemsByExtension: jest.fn(() => [{ route: 'some-route' }])
          })
          const wrapper = getWrapper(store)
          await wrapper.setData({ loadingGroups: false })
          const editRouteButton = wrapper.find(selectors.editRouteButton)
          expect(editRouteButton).toMatchSnapshot()
        })
      })
    })
  })

  describe('account information', () => {
    it('displays basic user information', () => {
      const store = getStore({
        user: {
          username: 'some-username',
          displayname: 'some-displayname',
          email: 'some-email'
        }
      })
      const wrapper = getWrapper(store)

      const accountPageInfo = wrapper.find(selectors.accountPageInfo)
      expect(accountPageInfo).toMatchSnapshot()
    })

    describe('group membership', () => {
      it('shows loading indicator when in loading state', () => {
        const wrapper = getWrapper()
        const groupInfoLoading = wrapper.find(selectors.loaderStub)
        expect(groupInfoLoading.exists()).toBeTruthy()
      })
      it('displays message if not member of any groups', async () => {
        const store = getStore()
        const wrapper = getWrapper(store)
        await wrapper.setData({ loadingGroups: false })

        const groupNamesEmpty = wrapper.find(selectors.groupNamesEmpty)
        expect(groupNamesEmpty.exists()).toBeTruthy()
      })
      it('displays group names', async () => {
        const store = getStore()
        const wrapper = getWrapper(store)
        await wrapper.setData({ loadingGroups: false })
        await wrapper.setData({ groups: ['one', 'two', 'three'] })

        const groupNames = wrapper.find(selectors.groupNames)
        expect(groupNames).toMatchSnapshot()
      })
    })
  })
})

function getWrapper(store = getStore()) {
  const component = {
    ...account,
    mounted: jest.fn()
  }
  const opts = {
    localVue,
    mocks: {
      $route
    },
    stubs: {
      'oc-spinner': true,
      'oc-button': true,
      'oc-icon': true
    },
    store
  }
  return shallowMount(component, opts)
}

function getStore({
  user = {},
  server = '',
  getNavItemsByExtension = jest.fn(() => []),
  isAccountEditingEnabled = true
} = {}) {
  return createStore(Vuex.Store, {
    getters: {
      user: () => user,
      configuration: () => ({
        server: server
      }),
      getNavItemsByExtension: () => getNavItemsByExtension,
      apps: () => ({
        ...(isAccountEditingEnabled || { settings: {} })
      })
    }
  })
}
