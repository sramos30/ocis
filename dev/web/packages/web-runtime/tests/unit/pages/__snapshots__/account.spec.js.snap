// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account page account information displays basic user information 1`] = `
<dl class="account-page-info oc-flex oc-flex-wrap">
  <div class="account-page-info-username oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Username</dt>
    <dd>
      some-username
    </dd>
  </div>
  <!---->
  <div class="account-page-info-displayname oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Display name</dt>
    <dd>
      some-displayname
    </dd>
  </div>
  <div class="account-page-info-email oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Email</dt>
    <dd>some-email</dd>
  </div>
  <div class="account-page-info-groups oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">
      Group memberships
    </dt>
    <dd data-testid="group-names">
      <oc-spinner-stub aria-label="Loading group membership information"></oc-spinner-stub>
    </dd>
  </div>
</dl>
`;

exports[`account page account information group membership displays group names 1`] = `<dd data-testid="group-names"><span>one, two, three</span></dd>`;

exports[`account page header section edit buttons edit route button should be displayed if running with ocis and has navItems 1`] = `
<oc-button-stub variation="primary" type="router-link" to="some-route" data-testid="account-page-edit-route-btn">
  <oc-icon-stub name="edit"></oc-icon-stub>
  <translate-stub tag="span">Edit</translate-stub>
</oc-button-stub>
`;

exports[`account page header section edit buttons edit url button should be displayed if not running with ocis 1`] = `
<oc-button-stub variation="primary" type="a" href="http://server/address/index.php/settings/personal" data-testid="account-page-edit-url-btn">
  <oc-icon-stub name="edit"></oc-icon-stub>
  <translate-stub tag="span">Edit</translate-stub>
</oc-button-stub>
`;
